src/test/scala/chiselTests/AnnotatingExample.scala

```diff
@@ -0,0 +1,157 @@
+}
+
+class AnnotatingExampleTester(a: Int, b: Int, z: Int) extends BasicTester {
+  val dut = Module(new AnnotatingExample)
+  val first = Reg(init=Bool(true))
+  dut.io.a := UInt(a)
+  dut.io.b := UInt(b)
+  dut.io.e := first
+  when(first) { first := Bool(false) }
+  when(!first && dut.io.v) {
+    assert(dut.io.z === UInt(z))
+    stop()
+  }
+}
+
+class AnnotatingExampleSpec extends FlatSpec with Matchers {
+  behavior of "Annotating components of a circuit"
+
+  val annotationMap = MyDriver.doStuff { () => new AnnotatingExampleTester(1, 2, 3) }
+  it should "contain the following relative keys" in {
+    annotationMap.contains("SomeSubMod.io.in") should be(true)
+    annotationMap("SomeSubMod.io.in") should be("sub mod io.in")
+  }
+  it should "contain the following absolute keys" in {
+    annotationMap.contains("AnnotatingExampleTester.dut.subModule2.io.out") should be (true)
+    annotationMap("AnnotatingExampleTester.dut.subModule2.io.out") should be ("sub mod io.out")
+  }
+}
+
+trait AnnotationScope
+case object AllRefs     extends AnnotationScope
+case object JustThisRef extends AnnotationScope
+
+object AnnotationKey {
+  def apply(component: SignalId): AnnotationKey = {
+    AnnotationKey(component, AllRefs)
+  }
+}
+case class AnnotationKey(val component: SignalId, scope: AnnotationScope) {
+  override def toString: String = {
+    scope match {
+      case JustThisRef =>
+        s"${component.pathName}"
+      case AllRefs =>
+        s"${component.parentModName}.${component.signalName}"
+      case  _ =>
+        s"${component.toString}_unknown_scope"
+    }
+  }
+}
+
+class AnnotationMap extends scala.collection.mutable.HashMap[AnnotationKey, String]
+
+class MyDynamicContext {
+  val annotationMap = new AnnotationMap
+}
+
+object MyBuilder {
+  private val myDynamicContextVar = new DynamicVariable[Option[MyDynamicContext]](None)
+
+  def myDynamicContext: MyDynamicContext =
+    myDynamicContextVar.value getOrElse new MyDynamicContext
+
+  def processAnnotations(annotationMap: AnnotationMap): Map[String, String] = {
+    val list = annotationMap.map { case (k,v) =>
+      k match {
+        case (AnnotationKey(signal, JustThisRef)) =>
+          f"Just this ref $k%60s -> $v%40s  component $signal"
+        case (AnnotationKey(signal, AllRefs)) =>
+          f"All refs      $k%60s -> $v%40s  component $signal"
+        case  _ =>
+          s"Unknown annotation key $k"
+      }
+    }.toList.sorted.mkString("\n")
+    println(s"Sorted list of annotations\n$list")
+
+    annotationMap.map { case (k,v) => k.toString -> v}.toMap
+  }
+
+  def build[T <: Module](f: => T): Map[String, String] = {
+    myDynamicContextVar.withValue(Some(new MyDynamicContext)) {
+      Driver.emit(() => f)
+      processAnnotations(myDynamicContextVar.value.get.annotationMap)
+    }
+  }
+}
+
+object MyDriver extends BackendCompilationUtilities {
+  def doStuff[T <: Module](gen: () => T): Map[String, String] = MyBuilder.build(Module(gen()))
```

Comment:
'doStuff' as a name is basically content-free. It doesn't give the reader any idea of what it does, any function could be named 'doStuff' and it would still be accurate.